content/general/concepts/what-is-the-prototype-pattern/what-is-the-prototype-pattern.md

```diff
@@ -0,0 +1,277 @@
+
+There are two types of copying associated with the prototype pattern. They are:
+
+- Shallow Copy (copies the immediate property values)
+- Deep Copy (copies the immediate values, plus any referenced object)
+
+## Example: Shallow Copy Vs Deep Copy
+
+To illustrate the difference between the shallow and deep copy, a model has been provided below, consisting of two classes `TvSeries` and `Episode`. Cloning functionality has been added to TvSeries by implementing the Cloneable interface.
+
+```java
+public class TvSeries implements Cloneable {
+  private final String name;
+  private final int seriesNo;
+  // Notice one or more properties are a complex type and mutable
+  private final List<Episode> episodes;
+
+  public TvSeries(String name, int seriesNo, List<Episode> episodes) {
+    this.name = name;
+    this.seriesNo = seriesNo;
+    this.episodes = episodes;
+  }
+
+  // Shallow copy method
+  @Override
+  public TvSeries clone() throws CloneNotSupportedException {
+    // Cast to TvSeries as super returns Object
+    return (TvSeries) super.clone();
+  }
+
+  // Deep copy
+  public TvSeries deepCopy() {
+    // New Episode objects are created during the copy
+    return new TvSeries(this.name, this.seriesNo, this.episodes.stream()
+                                                               .map(episode -> new Episode(episode.getName(), episode.getEpNo()))
+                                                               .collect(Collectors.toList()));
+  }
+
+  // Getters and toString
+}
+
+```
+
+```java
+public class Episode {
+  private String name;
+  private int epNo;
+
+  public Episode(String name, int epNo)
+  {
+    this.name = name;
+    this.epNo = epNo;
+  }
+
+  // Getters, setters and toString
+}
+```
+
+_Note:_ The properties of the `Episode` class are mutable. This is to illustrate a potential drawback in the shallow copy method implemented in the `TvSeries` class.
+
+The following code demonstrates a limitation of the shallow copy:
+
+```java
+public class Main {
+  public static void main(String[] args) throws CloneNotSupportedException {
+    // A 'new' TvSeries object is created as seriesOne
+    Episode episodeOne = new Episode("Hello Friend", 13);
+    final Series seriesOne = new Series("Mr Robot", 1, Collections.singletonList(episodeOne));
+
+    // A second TvSeries object is created by shallow copying seriesOne
+    final Series seriesTwo = seriesOne.clone();
+    // Change the name of S2ep1
+    seriesTwo.getEpisodes().get(0).setName("Unmask");
+    // S2ep1 name has changed
+    System.out.println(seriesTwo.getEpisodes().get(0).getName()); // Unmask
+    // But, so has S1ep1 :(
+    System.out.println(seriesOne.getEpisodes().get(0).getName()); // Unmask
```

Comment:
Breaking up the code snippets and with comments in between may seem like a good idea as we're documenting, but for readers, this can be a bit difficult to follow. Instead, these comments should be placed outside of the code snippet for better understanding.